src/Lean/Meta/Tactic/AC/Main.lean

```diff
@@ -312,19 +312,30 @@ partial def rewriteUnnormalized (mvarId : MVarId) : M Unit :=
         assignExprMVar mvarId proof
       else throwError ""
     catch _ => throwError "cannot synthesize proof:\n{MessageData.ofGoal mvarId}"
-  | ProofStrategy.rfl tgt =>
-    trace[Meta.AC] "picking rfl strategy {MessageData.ofGoal mvarId}"
-    assignExprMVar mvarId (←mkAppM ``Eq.refl #[tgt])
+  | ProofStrategy.simp =>
+    trace[Meta.AC] "picking simp strategy {MessageData.ofGoal mvarId}"
+    let simpCtx ← Simp.Context.mkDefault
+    let newGoal ← simpTarget mvarId simpCtx
+    unless newGoal.isNone do
+      throwError "cannot synthesize proof:\n{MessageData.ofGoal mvarId}"
   | ProofStrategy.norm (definitelyNormalized _) => throwError "no unnormalized operators found"
   | ProofStrategy.norm e =>
     trace[Meta.AC] "picking norm strategy {MessageData.ofGoal mvarId}"
     let mvarId ← addAcEq mvarId e target
     let (h_ac, mvarId) ← intro mvarId `h_ac
-    let res ← rewrite mvarId (←getMVarType mvarId) (mkFVar h_ac)
-    let [] := res.mvarIds | throwError "no meta variables expected after rewrite"
-    let mvarId ← replaceTargetEq mvarId res.eNew res.eqProof
-    let mvarId ← clear mvarId h_ac
-    rewriteUnnormalized mvarId
+    let simpCtx ← Simp.Context.mkDefault
+    withMVarContext mvarId do
+    let simpCtx := { simpCtx with simpTheorems := ←simpCtx.simpTheorems.add #[] (mkFVar h_ac) }
+
+    trace[Meta.AC] "pre rewrite state:\n{MessageData.ofGoal mvarId}\n"
+    let mvarId ← simpTarget mvarId simpCtx
```

Comment:
I think `Simp.main` function is more appropriate. It allows us to provide two functions `pre` and `post` that are invoked before/after we visit a term. We should invoke the ac normalization code in the `post`. 
Here is an example for `Simp.main`
https://github.com/leanprover/lean4/blob/master/src/Lean/Meta/Tactic/Split.lean#L19-L33

Reply:
Here is an example where we check the "parent" term to decide whether we are going to "simplify" something or not.
https://github.com/leanprover/lean4/blob/master/src/Lean/Meta/Tactic/LinearArith/Simp.lean#L16-L24
You have to do something similar when invoking the AC normalizer in the post function. If the parent is the same operator, we should not normalize.

Reply:
done